ip_proofs/src/gipa.rs

```diff
@@ -0,0 +1,556 @@
+{
+    pub(crate) r_commitment_steps: Vec<(
+        (LMC::Output, RMC::Output, IPC::Output),
+        (LMC::Output, RMC::Output, IPC::Output),
+    )>,
+    pub(crate) r_base: (LMC::Message, RMC::Message),
+    _gipa: PhantomData<GIPA<IP, LMC, RMC, IPC, D>>,
+}
+
+#[derive(Clone)]
+pub struct GIPAAux<IP, LMC, RMC, IPC, D>
+where
+    D: Digest,
+    IP: InnerProduct<
+        LeftMessage = LMC::Message,
+        RightMessage = RMC::Message,
+        Output = IPC::Message,
+    >,
+    LMC: DoublyHomomorphicCommitment,
+    RMC: DoublyHomomorphicCommitment<Scalar = LMC::Scalar>,
+    IPC: DoublyHomomorphicCommitment<Scalar = LMC::Scalar>,
+    RMC::Message: MulAssign<LMC::Scalar>,
+    IPC::Message: MulAssign<LMC::Scalar>,
+    RMC::Key: MulAssign<LMC::Scalar>,
+    IPC::Key: MulAssign<LMC::Scalar>,
+    RMC::Output: MulAssign<LMC::Scalar>,
+    IPC::Output: MulAssign<LMC::Scalar>,
+{
+    pub(crate) r_transcript: Vec<LMC::Scalar>,
+    pub(crate) ck_base: (LMC::Key, RMC::Key),
+    _gipa: PhantomData<GIPA<IP, LMC, RMC, IPC, D>>,
+}
+
+//TODO: Can extend GIPA to support "identity commitments" in addition to "compact commitments", i.e. for SIPP
+
+impl<IP, LMC, RMC, IPC, D> GIPA<IP, LMC, RMC, IPC, D>
+where
+    D: Digest,
+    IP: InnerProduct<
+        LeftMessage = LMC::Message,
+        RightMessage = RMC::Message,
+        Output = IPC::Message,
+    >,
+    LMC: DoublyHomomorphicCommitment,
+    RMC: DoublyHomomorphicCommitment<Scalar = LMC::Scalar>,
+    IPC: DoublyHomomorphicCommitment<Scalar = LMC::Scalar>,
+    RMC::Message: MulAssign<LMC::Scalar>,
+    IPC::Message: MulAssign<LMC::Scalar>,
+    RMC::Key: MulAssign<LMC::Scalar>,
+    IPC::Key: MulAssign<LMC::Scalar>,
+    RMC::Output: MulAssign<LMC::Scalar>,
+    IPC::Output: MulAssign<LMC::Scalar>,
+{
+    pub fn setup<R: Rng>(
+        rng: &mut R,
+        size: usize,
+    ) -> Result<(Vec<LMC::Key>, Vec<RMC::Key>, IPC::Key), Error> {
+        Ok((
+            LMC::setup(rng, size)?,
+            RMC::setup(rng, size)?,
+            IPC::setup(rng, 1)?.pop().unwrap(),
+        ))
+    }
+
+    pub fn prove(
+        values: (&[IP::LeftMessage], &[IP::RightMessage], &IP::Output),
+        ck: (&[LMC::Key], &[RMC::Key], &IPC::Key),
+        com: (&LMC::Output, &RMC::Output, &IPC::Output),
+    ) -> Result<GIPAProof<IP, LMC, RMC, IPC, D>, Error> {
+        if IP::inner_product(values.0, values.1)? != values.2.clone() {
+            return Err(Box::new(InnerProductArgumentError::InnerProductInvalid));
+        }
+        if values.0.len().count_ones() != 1 {
+            // Power of 2 length
+            return Err(Box::new(InnerProductArgumentError::MessageLengthInvalid(
+                values.0.len(),
+                values.1.len(),
+            )));
+        }
+        if !(LMC::verify(ck.0, values.0, com.0)?
+            && RMC::verify(ck.1, values.1, com.1)?
+            && IPC::verify(&vec![ck.2.clone()], &vec![values.2.clone()], com.2)?)
```

Comment:
should we parallelize these? Or do you think it would be unnecessary (verification is already fast enough)?

Reply:
Yes, verification will be a significant cost. I think instead of parallelizing the calls to each of these three verifies, we will get what we want when we implement parallelism for the inner products in `inner_products`, e.g. `pedersen` uses `MultiexponentiationInnerProduct` and `afgho16` uses `PairingInnerProduct`.